File: julia/01/main.jl
function readip(fn)
    open(fn) do f
        readlines(f)
    end
end

function numsonly(s,nums)
    result = ""
    for i = 1:length(s)
        subs = s[i:end]
        for (s1,s2) in nums
            if startswith(subs,s1)
                result *= s2
            end
        end
    end
    result
end

nums1 = begin
    d = Dict{String,String}()
    for i in 0:9
        d[string(i)] = string(i)
    end
    d
end

nums2 = begin
    d = copy(nums1)
    d["one"] = "1"
    d["two"] = "2"
    d["three"] = "3"
    d["four"] = "4"
    d["five"] = "5"
    d["six"] = "6"
    d["seven"] = "7"
    d["eight"] = "8"
    d["nine"] = "9"
    d["zero"] = "0"
    d
end

function part1(fn)
    ip = readip(fn)
    snos = map((s)->numsonly(s,nums1),ip)
    nos = map((s)->parse(Int64,s[1]*s[end]),snos)
    println("part 1 ",fn, " ans = ",sum(nos))
end


function part2(fn)
    ip = readip(fn)
    snos = map((s)->numsonly(s,nums2),ip)
    nos = map((s)->10*parse(Int64,s[1]) + parse(Int64,s[end]),snos)
    println("part 2 ",fn, " ans = ",sum(nos))
end

part1("test_input")
part1("input")
part2("test_input2")
part2("input")

File: cpp/01/main.cpp
#include "aoc.h"

using namespace std;

string only_digits(string s) {
    string op;
    for(char c : s) {
        if(c >= '0' && c <= '9') op.push_back(c);
    }
    return op;
}

string only_nums(string s) {
    map<string,string> nums = {
        {"zero","0"},
        {"one","1"},
        {"two","2"},
        {"three","3"},
        {"four","4"},
        {"five","5"},
        {"six","6"},
        {"seven","7"},
        {"eight","8"},
        {"nine","9"}
    };

    for(int i = 0; i < 10; i++) {
        string s = to_string(i);
        nums.insert({s,s});
    }

    string op;
    for(size_t i = 0; i < s.size(); i++) {
        string subs = s.substr(i);
        for(auto [s1,s2] : nums) {
            if(subs.starts_with(s1)) {
                op += s2;
                break;
            }
        }
    }
    return op;
}

void part1(const string & fn) {
    auto ip = read_ip(fn);
    size_t ans = 0;
    for(auto line : ip) {
        string s1 = only_digits(line);
        string s2;
        s2.push_back(s1.front());
        s2.push_back(s1.back());
        size_t n = stoi(s2);
        ans += n;
    }
    cout << fn << " day 1 part 1 answer = " << ans << endl;
}

void part2(const string & fn) {
    auto ip = read_ip(fn);
    size_t ans = 0;
    for(auto line : ip) {
        string s1 = only_nums(line);
        string s2;
        s2.push_back(s1.front());
        s2.push_back(s1.back());
        size_t n = stoi(s2);
        ans += n;
    }
    cout << fn << " day 1 part 2 answer = " << ans << endl;
}

int main() {
    //part1("test_input");
    part1("input");
    //part2("test_input2");
    part2("input");
    return 0;
}